core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -156,6 +161,52 @@ protected void perform(RelOptRuleCall call, Filter filter,
     final List<RexNode> leftFilters = new ArrayList<>();
     final List<RexNode> rightFilters = new ArrayList<>();
 
+    final int nFieldsLeft = join
+        .getInputs()
+        .get(0)
+        .getRowType()
+        .getFieldList()
+        .size();
+    final int nFieldsRight = join
+        .getInputs()
+        .get(1)
+        .getRowType()
+        .getFieldList()
+        .size();
+
+    final int nTotalFields = join.getRowType().getFieldList().size();
+    final int nSysFields = 0; // joinRel.getSystemFieldList().size();
+
+    // Setup two bit sets, one for the left side of the join indicating what fields are present
+    // in the left side of the join and the other for the right side.
+    final ImmutableBitSet leftBitmap = ImmutableBitSet.range(
+        nSysFields, nSysFields + nFieldsLeft);
+    final ImmutableBitSet rightBitmap = ImmutableBitSet.range(
+        nSysFields + nFieldsLeft, nTotalFields);
+
+    // When a join can create nulls on the left (or right) leverage the filter to determine if
+    // the join can cancel nulls on left (or right).
+    if (joinType.generatesNullsOnLeft() && filter != null) {
+      final RexVisitor<Boolean> mustMatchSideVisitor = new TableNotNullable(leftBitmap);
+      final Boolean filterMustMatch = filter.getCondition().accept(mustMatchSideVisitor);
+      if (filterMustMatch != null) {
+        if (filterMustMatch != Boolean.FALSE) {
+          joinType = joinType.cancelNullsOnLeft();
+        }
+      }
+    }
+
+    // When a join can create nulls on the right (or left) leverage the filter to determine if
+    // the join can cancel nulls on right (or left).
+    if (joinType.generatesNullsOnRight() && filter != null) {
+      final RexVisitor<Boolean> mustMatchSideVisitor = new TableNotNullable(rightBitmap);
+      final Boolean filterMustMatch = filter.getCondition().accept(mustMatchSideVisitor);
+      if (filterMustMatch != null) {
+        if (filterMustMatch != Boolean.FALSE) {
```

Comment:
When the `filterMustMatch ` can be null ?

Reply:
`TableNotNullable` extends `RexVisitorImpl` and `RexVisitorImpl` returns `null` for all `visitXXXX` calls. This is convenient -- it allows classes that extend `RexVisitorImpl` to only implement the `visitXXX` methods that are relevant.

I have been tempted to update `RexVisitorImpl` to have a constructor like so:
```java
RexVisitorImpl(boolean deep, R defaultValue) {
  this.deep = deep;
  this.defaultValue = defaultValue;
}
```
This would prevent null the need to handle `null` everywhere.

Reply:
I wouldn't put too much effort into this visitor. Before this PR is accepted I would like an independent utility, defined outside of the rule, with lots of unit tests. As I have said, I think `Strong` can be extended to meet that need. In any case, the unit tests are going to be as comprehensive as `RexProgramTest.testStrong`, which contains about 50 test cases.

Reply:
Agreed this is fairly naive. Could you share two cases you expect to be troublesome? Be a good jumping off point for me. 
"Sub query with filter" 
ANY clause.
IF NOT DISTINCT?

Reply:
I don't know what would be troublesome. I do know what we don't know what we don't know. And therefore starting from scratch is a mistake.

A couple of examples: 
* '(x IS NOT TRUE) IS NOT NULL' (the IS functions convert 3vl to 2vl, so, like NOT they have complex truth tables)
* '(CASE WHEN x = 1 THEN 5 ELSE y END) > z' (you can isolate the expression to left or right input only if you push '> z' into the CASE)
* and any expression that is deeply nested (e.g. contains NOT inside CASE inside NOT)